keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2488 @@
+Kubernetes API server as an alternative to validating admission webhooks.
+
+This proposal builds on the capabilities of the [CRD Validation
+Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+feature that graduated to beta in 1.25, but with a focus on the policy
+enforcement capabilities of validating admission control.
+
+## Motivation
+
+This KEP will lower the infrastructure barrier to enforcing customizable
+policies as well as providing primitives that help the community establish and
+adhere to best practices of both K8s and its extensions.
+
+Currently the way custom policies are enforced are via admission webhooks.
+Admission webhooks are extremely flexible, but have a few drawbacks as compared
+to in-process policy enforcement:
+
+- They require building infrastructure to host the admission webhook.
+- They contribute to latency by requiring another network hop.
+- Due to the extra infrastructure dependencies, webhooks are inherently less
+  reliable than in-process webhooks. This forces cluster operators to choose
+  between failing closed, which reduces the availability of the cluster as a
+  whole and failing open, which limits the efficacy of webhooks for enforcing
+  policy.
+- Webhooks are operationally burdensome for cluster administrators to
+  manage. They must take responsibility for the observability, security and the
+  release/rollout/rollback plans for the webhook.
+
+Taking a view of the K8s ecosystem as a whole, it is clear that there is demand
+for opinionated policy frameworks.
+
+Pod Security Policies provided this for pods, but encountered a number of
+issues. One of which was that it was hard to keep up with community demand for
+more control surfaces, and the delay in delivering these control surfaces due to
+K8s' rollout period.
+
+Pod Security Admission is a similar solution, but does not attempt to duplicate
+the control granularity that PSP provided.
+
+There are numerous in-tree embedded controllers.
+
+The existence of security regimes like the CIS Kubernetes Benchmarks highlight
+the values of standardized controls. Automating their enforcement, where
+possible, will make it easier for users to lock down their clusters.
+
+With the advent of CRDs, and the drive to make the resources they define
+first-class entities, the footprint of Kubernetes extensions is set to grow for
+the foreseeable future. This KEP allows authors of such extensions to provide
+policy primitives similar to PSP or PSA, putting them on equal footing with
+in-tree functionality.
+
+With the reduced infrastructure footprint and demonstrated demand for a
+customizable, built-in mechanism for extensible policy, this KEP fills a
+community need. It is not intended to replace validating admission webhooks
+altogether, however, since these can support functionality that may not make
+sense to provide in-tree.
+
+### Goals
```

Comment:
sig-auth has previous talked about trying to find a way to restrict access from a daemonset pod to a customresource/foo that has Foo.spec.NodeName set to the Node.metadata.name of the pod bound to the particular SA token. This is tantalizingly close because user.Extra contains `authentication.kubernetes.io/pod-uid` to locate a pod, determine a Pod.spec.NodeName.

A built-in that does that may be well received and unlock many use-cases.  Exploring the idea may be useful.  If most also require controlled read permission, then its probably better to create something specifically for the purpose.

Reply:
I've added a note to the KEP. Sounds like we will need to investigate. Looking up the pod (or any other additional resources) is not something we were planning to allow, but I'd still like to understand this one better. I'll bring it up with sig-auth.